### DISCUSSION TITLE
Android Permission Error

### DISCUSSION THREAD

**Discussion by glovas1** [beeware/toga#1468] [Q&A]
I am new to Python, Beeware and Toga. I apologize if this is not the right forum to pose this question.

I have an new app that is working in Windows and I am trying to get it to run on Android so I can take it along with me. It is to read a directory listing file of my iTunes pulled off my media server. I can then use the mobile to see if I have the song/CD already when I come across an album that I may not have.

I import os, use a call to os.getcwd() and it has a return (but I suspect it is just showing me the root of the application??). When I do another call to load os.listdir() to a list, I get the permissions error below. Any help would be most appreciated!!

04-22 16:15:05.134  4507  4507 E Python  :     me = os.listdir()
04-22 16:15:05.134  4507  4507 E Python  : PermissionError: [Errno 13] Permission denied

**Comment by freakboy3742** [comment]
You've definitely found the right place - welcome!

The problem you're hitting is something common on mobile platforms - they're a lot more locked down than desktop platforms in terms of what you're allowed to see. This is part of the "app sandbox" - one of the reasons that apps on phones are more secure is that they very strictly regulate what you can and can't see.

In this case, the current working directory (which is a bit of an odd concept on a phone to start with) exists, but it isn't guaranteed to be a location that you can view.

The good news is that Toga apps expose a collection of root paths that *are* visible. Inside your Toga App object, there is an `paths` attribute - this attribute has a bunch of sub attributes (`app`, `data`, `cache` etc) that are folder locations that are known to exist, and known to be user-readable. The `app` path is the location of your app's code, so if you're bundling resources with your app, that's the base path you should use. These attributes are all `pathlib.Path` objects, so you won't need to use the os module at all.

**Reply by glovas1** [reply]
Thank you for the quick answer. I will search for documentation to see where and how to apply this to my app. I am retired so have spent about 2 weeks on and off trying to find a solution.  :-)

**Reply by emendir** [reply]
Can't one give files in any of the mentioned directories execute permission?
I tried with:
```python
st = os.stat(path_to_binary)
os.chmod(path_to_binary, st.st_mode | stat.S_IEXEC)
```
but I still get a `Permission denied` error when trying to execute it with `subprocess.Popen`

**Reply by mhsmith** [reply]
Unfortunately, Android API level 29 and higher uses SELinux to block apps from running executables in their data directory. It may be possible to work around this by putting them in the APK libs directory and using `extractNativeLibs`:
* https://issuetracker.google.com/issues/128554619?pli=1#comment4
* https://android-review.googlesource.com/c/platform/system/sepolicy/+/804149

**Reply by freakboy3742** [reply]
Mobile platforms have made it very clear that they don't want you doing this sort of thing - Android isn't just a "portable linux server", so you shouldn't assume you can fork/spawn to a subprocess at will. 

Even if you're are able to make this work as @mhsmith suggests, I'd strongly advise against it. You can reasonably expect that the operating system will become increasingly hostile towards this sort of capability over time. Any app that relies on invoking subprocesses for workflows is going to be fighting a losing battle against operating system constraints every time there is an OS update. It's completely prohibited on iOS (and has been since day 1); and Android has been trending increasingly in this direction over time.

Also - you haven't described where you obtained this file that you want to be executable, but just in case: something you may not have considered is that you need the executable to be an *Android* executable, not just something that runs on your own machine. Just copying an executable file from your development machine won't work.
